mpip: add v3.5 release

cchambreau opened the pull request:
Updates for 3.5 release as well as 3.4.1 release.

https://software.llnl.gov/mpiP/

cchambreau committed:
Update for mpip-3.5 release

adamjstewart commented on var/spack/repos/builtin/packages/mpip/package.py:
```diff
@@ -4,29 +4,121 @@
+
+    conflicts('platform=darwin')
+
+    depends_on('python')
+    depends_on('python@:2', when='@:3.4.1')
```
What kind of dependency is this? build? run? link?

adamjstewart commented on var/spack/repos/builtin/packages/mpip/package.py:
```diff
@@ -4,29 +4,121 @@
+    variant('setjmp', default=False,
+            description="Use setjmp to generate stack trace")
+
+    variant('mpi-io', default=True,
```
I try to avoid variants with dashes in the name since dashes are also used to deactivate variants. Maybe an underscore would be better?

adamjstewart commented on var/spack/repos/builtin/packages/mpip/package.py:
```diff
@@ -4,29 +4,121 @@
+            config_args.append('--disable-libunwind')
+            config_args.append('--enable-setjmp')
+
+        maxargs = int(spec.variants['maxargs'].value)
+        if maxargs != 32:
+            config_args.extend(['--enable-maxargs={0}'.format(maxargs)])
+
+        stackdepth = int(spec.variants['stackdepth'].value)
+        if stackdepth != 8:
+            config_args.extend(['--enable-stackdepth={0}'.format(stackdepth)])
+
+        internal_stackdepth = int(spec.variants['internal-stackdepth'].value)
+        if internal_stackdepth != 3:
+            config_args.extend(['--enable-internal-stackdepth={0}'
+                               .format(internal_stackdepth)])
```
I think it's better to pass the default, since the default may change in the future.

adamjstewart commented on var/spack/repos/builtin/packages/mpip/package.py:
```diff
@@ -4,29 +4,121 @@
+    #    setjmp functionality, if needed
+    #  depends_on('unwind')
+
+    variant('demangling', default=True,
```
I usually put variants before dependencies and conflicts.

lee218llnl commented:
There are a few when='@:3.4.1' conditions that don't appear to be necessary. I only see versions 3.4.1 and 3.5 in the spack package.py file and on the mpip github releases page.

cchambreau committed:
Additional changes to support 3.5 and 3.4.1 releases

cchambreau commented:
I believe I have addressed all the comments.  

- Python is required for generating the MPI wrappers.  I added a comment regarding this and believe that I have clarified the python dependencies.  

- I renamed the older 3.4.1 variant "shared" to "add_shared_target" to hopefully reduce confusion with releases after 3.4.1, since the default behavior with 3.5 is to build the shared library.

adamjstewart commented on var/spack/repos/builtin/packages/mpip/package.py:
```diff
@@ -4,37 +4,127 @@
+
+    conflicts('platform=darwin')
+
+    # make-wrappers.py wrapper generator script requres python
```
```suggestion
    # make-wrappers.py wrapper generator script requires python
```
Is this at build-time or at run-time? It sounds like you don't link to the Python libraries, so the current deptype is incorrect.

cchambreau committed:
Specify python deptype
